packages/x-date-pickers-pro/src/internal/utils/date-utils.ts

```diff
@@ -13,12 +13,20 @@ export const isRangeValid = <TDate>(
   return Boolean(range && range[0] && range[1] && !utils.isBefore(range[1], range[0]));
 };
 
-export const isWithinRange = <TDate>(
+export const isWithinDayRange = <TDate>(
   utils: MuiPickersAdapter<TDate>,
   day: TDate,
   range: DateRange<TDate> | null,
 ) => {
-  return isRangeValid(utils, range) && utils.isWithinRange(day, range);
+  if (!isRangeValid(utils, range)) {
+    return false;
+  }
+
+  const cleanRange = [
```

Comment:
It seems strange that we'd need such data manipulation on every day comparison. 🤔 
I've tried debugging this method both on master and next branches with the following breakpoint:
`!!range && range[0] && range[0].hour() !== 0` and discovered that it is being triggered only on `next`, whereas on `master` this method receives day without hours, even though the defined value has hours (same demo used).
It seems that the root cause of the regression is somewhere up the chain. 🤔 

Are we sure that we want to fix it with such somewhat expensive (day manipulation) operation?

Reply:
The change likely come from `parseRangeInputValue` which was moving to the start of the day.
I think we do have to support date with hours deeper in the tree now.
Because of #6416 (which has no parsing) and because of the future `DateTimeRangePicker` where we will have to keep the time.

Reply:
But I could probably do this transformation inside the component to avoid doing it for every day

Reply:
Gotcha.
Well, in such case it would still be nicer to user utility methods to overcome this issue instead of date manipulation.
Something along these lines:
```diff
- day.isBetween(range[0], range[1], null, '[]');
+ day.isBetween(range[0], range[1], 'day', '[]');
```
But that is not supported with current date-io implementation and I haven't checked if all date-io supported libraries do support `granularity` option.

Reply:
> But I could probably do this transformation inside the component to avoid doing it for every day

Yes, my concern is the amount of additional date modifications we are going to do after this change 🙈

Reply:
I pushed a commit to do the transform in the render.

> Well, in such case it would still be nicer to use utility methods to overcome this issue instead of date manipulation.

I agree that it would be a cleaner long term solution.

Reply:
> I haven't checked if all date-io supported libraries do support granularity option.

I've checked date-io and can conclude that this feature would be tricky to implement as only a handful of date-util libraries do support comparison granularity. 😢